{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import openpyxl as op\n",
    "import numpy as np\n",
    "import scipy.misc as smp\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Makes heatmaps from 3 sources of data: PHASTER, MvirDB blast, and Plasmid Finder\n",
    "\n",
    "Different classes - go by columns.\n",
    "\n",
    "Different rows - different taxa.\n",
    "\n",
    "For this pass, we'll just classify everything, put some white spaces in-between columns.\n",
    "When we get lists of different clades, we can make them different colors, potentially.\n",
    "\n",
    "Some alternate color scheme should be added later."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "##Tree order\n",
    "with open('./../RAxML Trees/RAxML_bipartitions.RAxML-concat-ccg_2-nd.nex', 'r') as f:\n",
    "    nexus = f.read()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "tree = nexus.split('begin trees;')[1].split('end;')[0].strip()\n",
    "splits = ['S'+i.split(':')[0] for i in tree.split('S')[1:]]\n",
    "for i in range(len(splits)):\n",
    "    if ':' in splits[i]:\n",
    "        splits[i] = splits[i].split(':')[0]\n",
    "go = splits\n",
    "\n",
    "gos = ['SEQ000010060','SAMN02844600','SEQ000008922','SEQ000004791','SAMN02844917',\\\n",
    "      'SEQ000011996','SEQ000019495','SEQ000015224','SEQ000014214','SEQ000012954',\\\n",
    "      'SEQ000034510','SEQ000037525']\n",
    "\n",
    "with open('2017-05-31_tree_order.txt','w') as f:\n",
    "    f.write('\\n'.join(go))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SEQ000007959 Salmon_RE_2010_NC_019488 dupe\n",
      "SEQ000015055 Salmon_Fels_2_NC_010463 dupe\n",
      "SEQ000030500 Phage_Gifsy_1_NC_010392 dupe\n",
      "SEQ000033581 Phage_Gifsy_1_NC_010392 dupe\n",
      "SEQ000038289 Salmon_SSU5_NC_018843 dupe\n",
      "SEQ000038380 Salmon_Fels_2_NC_010463 dupe\n",
      "SEQ000038382 Salmon_Fels_2_NC_010463 dupe\n",
      "SEQ000039232 Salmon_Fels_2_NC_010463 dupe\n",
      "SEQ000039242 Salmon_Fels_2_NC_010463 dupe\n",
      "intact viruses found: 1122\n",
      "phages total: 586 near ends: 144\n",
      "Entero_186_NC_001317 23 2\n",
      "Entero_P88_NC_026014 12 1\n",
      "Entero_PsP3_NC_005340 20 1\n",
      "Escher_D108_NC_013594 13 1\n",
      "Haemop_HP2_NC_003315 18 3\n",
      "Phage_Gifsy_1_NC_010392 102 36\n",
      "Phage_Gifsy_2_NC_010393 47 14\n",
      "Salmon_118970_sal3_NC_031940 37 16\n",
      "Salmon_Fels_2_NC_010463 95 10\n",
      "Salmon_RE_2010_NC_019488 20 2\n",
      "Salmon_SEN34_NC_028699 50 18\n",
      "Salmon_SP_004_NC_021774 21 0\n",
      "458 104\n",
      "46 different viruses putatively identified\n",
      "['Edward_GF_2_NC_026611', 'Entero_186_NC_001317', 'Entero_Arya_NC_031048', 'Entero_ES18_NC_006949', 'Entero_Mu_NC_000929', 'Entero_P1_NC_005856', 'Entero_P22_NC_002371', 'Entero_P4_NC_001609', 'Entero_P88_NC_026014', 'Entero_PsP3_NC_005340', 'Entero_SfV_NC_003444', 'Entero_Tyrion_NC_031077', 'Entero_UAB_Phi20_NC_031019', 'Entero_fiAA91_ss_NC_022750', 'Entero_lambda_NC_001416', 'Entero_mEp460_NC_019716', 'Escher_D108_NC_013594', 'Escher_phiV10_NC_007804', 'Escher_pro483_NC_028943', 'Haemop_HP2_NC_003315', 'Klebsi_phiKO2_NC_005857', 'Pasteu_F108_NC_008193', 'Phage_Gifsy_1_NC_010392', 'Phage_Gifsy_2_NC_010393', 'Pseudo_MD8_NC_031091', 'Salmon_103203_sal5_NC_031946', 'Salmon_118970_sal3_NC_031940', 'Salmon_118970_sal4_NC_030919', 'Salmon_Fels_1_NC_010391', 'Salmon_Fels_2_NC_010463', 'Salmon_RE_2010_NC_019488', 'Salmon_SEN1_NC_029003', 'Salmon_SEN34_NC_028699', 'Salmon_SEN4_NC_029015', 'Salmon_SEN5_NC_028701', 'Salmon_SJ46_NC_031129', 'Salmon_SPN1S_NC_016761', 'Salmon_SPN3UB_NC_019545', 'Salmon_SP_004_NC_021774', 'Salmon_SP_031_NC_021775', 'Salmon_SSU5_NC_018843', 'Salmon_ST64B_NC_004313', 'Salmon_epsilon15_NC_004775', 'Salmon_vB_SosS_Oslo_NC_018279', 'Shigel_SfII_NC_021857', 'Vibrio_X29_NC_024369']\n"
     ]
    }
   ],
   "source": [
    "##PHASTER OUTPUT##\n",
    "\n",
    "#cutoff for phage calling\n",
    "cutoff = True\n",
    "co = 50 #% of genes of query region found in virus\n",
    "end_tolerance = 100 #amount of tolerance to be considered close to a contig end\n",
    "#gives some leeway in colling CDSs\n",
    "\n",
    "viruses_near_ends = []\n",
    "intact_viruses = 0\n",
    "\n",
    "##first, scrape the data\n",
    "phages = []\n",
    "phd = {} #phage dictionary\n",
    "phasterd = '/Users/jay.worley/Desktop/PHASTER/unzipped_results/'\n",
    "for i in [i for i in os.listdir(phasterd) if os.path.isdir(phasterd+i) and i in go]:\n",
    "    resultsf = phasterd+i+'/summary.txt'\n",
    "    with open(resultsf, 'r') as f:\n",
    "        rawresults = f.read()\n",
    "    results = rawresults.split('\\n\\n\\n\\n\\n')[1].strip().split('\\n')[3:]\n",
    "    \n",
    "    for result in results:\n",
    "        rl = ([i.strip() for i in result.split(' ') if len(i)>0])\n",
    "        if rl[2].startswith('intact'):\n",
    "            genome = rl[4].split('_')[0]\n",
    "            phage = '_'.join(rl[13].split('(')[0].split('_')[1:])\n",
    "            pct = float(rl[15].split('%')[0])\n",
    "            \n",
    "            intact_viruses += 1\n",
    "            #cutoff if called for\n",
    "            if cutoff == True and pct < co:\n",
    "                continue\n",
    "            #detect hits split by ends of contig or near end of contig\n",
    "            contig_len = int(rl[4].split('_')[4])\n",
    "            hit_start = int(rl[4].split(':')[1].split('-')[0])\n",
    "            hit_stop = int(rl[4].split(':')[1].split('-')[1])\n",
    "            \n",
    "            if any([end_tolerance>hit_start, end_tolerance>hit_stop, (contig_len-end_tolerance)<hit_start, (contig_len-end_tolerance)<hit_stop]):\n",
    "                viruses_near_ends.append(phage)\n",
    "                \n",
    "            phages.append(phage)\n",
    "            if genome not in phd.keys():\n",
    "                phd[genome] = {}\n",
    "            if phage not in phd[genome].keys():\n",
    "                phd[genome][phage] = pct\n",
    "                continue\n",
    "            elif pct > phd[genome][phage]:\n",
    "                phd[genome][phage] = pct\n",
    "                print(genome,phage,'dupe')\n",
    "            else:\n",
    "                print(genome,phage,'dupe')\n",
    "                continue\n",
    "            continue\n",
    "print('intact viruses found:', intact_viruses)            \n",
    "print('phages total:',len(phages),'near ends:',len(viruses_near_ends))\n",
    "classes = sorted(set(phages))\n",
    "tot_clas = []\n",
    "tot_ends = []\n",
    "for clas in classes:\n",
    "    if phages.count(clas) > 10:\n",
    "        print(clas, phages.count(clas), viruses_near_ends.count(clas))\n",
    "        tot_clas.append(int(phages.count(clas)))\n",
    "        tot_ends.append(int(viruses_near_ends.count(clas)))\n",
    "print(sum(tot_clas), sum(tot_ends))\n",
    "phages = sorted(set(phages))\n",
    "print(len(phages), 'different viruses putatively identified')\n",
    "print(phages)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "577\n",
      "320\n",
      "6\n"
     ]
    }
   ],
   "source": [
    "num_viruses = []\n",
    "for i in phd:\n",
    "    num_viruses.append(len(phd[i]))\n",
    "print(sum(num_viruses))\n",
    "print(len(phd))\n",
    "print(max(num_viruses))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 Entero_186_NC_001317            count: 23 \tproportion: 0.051685393258426963\n",
      "2 Entero_P88_NC_026014            count: 12 \tproportion: 0.02696629213483146\n",
      "3 Entero_PsP3_NC_005340           count: 20 \tproportion: 0.0449438202247191\n",
      "4 Escher_D108_NC_013594           count: 13 \tproportion: 0.029213483146067417\n",
      "5 Haemop_HP2_NC_003315            count: 18 \tproportion: 0.04044943820224719\n",
      "6 Phage_Gifsy_1_NC_010392         count: 100 \tproportion: 0.2247191011235955\n",
      "7 Phage_Gifsy_2_NC_010393         count: 47 \tproportion: 0.10561797752808989\n",
      "8 Salmon_118970_sal3_NC_031940    count: 37 \tproportion: 0.08314606741573034\n",
      "9 Salmon_Fels_2_NC_010463         count: 90 \tproportion: 0.20224719101123595\n",
      "10 Salmon_RE_2010_NC_019488        count: 19 \tproportion: 0.04269662921348315\n",
      "11 Salmon_SEN34_NC_028699          count: 50 \tproportion: 0.11235955056179775\n",
      "12 Salmon_SP_004_NC_021774         count: 21 \tproportion: 0.04719101123595506\n",
      "Min: 0 \tMax: 5 \tAvg: 1.0112359550561798\n"
     ]
    }
   ],
   "source": [
    "##NOW CREATE THE IMAGE##\n",
    "\n",
    "#phage count cutoff\n",
    "\n",
    "cutoff = True\n",
    "if cutoff == True:\n",
    "    co = 10 #cutoff value\n",
    "    phagedump = []\n",
    "    for genome in phd.keys():\n",
    "        phagedump = phagedump + (list(phd[genome].keys()))\n",
    "    phages = [i for i in phages if phagedump.count(i) > co]\n",
    "    \n",
    "##COLOR VALUES\n",
    "#set what percent value should be red.\n",
    "#will be linearly r->g up to 100%\n",
    "redpct = 50\n",
    "\n",
    "\n",
    "phc = {} #Phage counts\n",
    "gphc = {}\n",
    "#create a matrix of the image size\n",
    "#((row, col), pct) -> ((genome,phage),match)\n",
    "\n",
    "##pixel dimensions\n",
    "bh = 10 #box height\n",
    "bw = 40 #box width\n",
    "lw = 3 #line width\n",
    "ow = 5 #outline width\n",
    "black = [0,0,0]\n",
    "white = [255,255,255]\n",
    "grey = [100,100,100]\n",
    "\n",
    "height = (len(go)*(bh+lw))-lw+(2*ow)\n",
    "width = (len(phages)*(bw+lw))-lw+(2*ow)\n",
    "\n",
    "data = np.zeros((height,width,3), np.uint8)\n",
    "\n",
    "##color lines white\n",
    "\n",
    "color_my_lines = True\n",
    "if color_my_lines == True:\n",
    "    color = black\n",
    "    lxs = [range(ow-lw+((bw+lw)*i),(ow+((bw+lw)*i))) for i in range(1,len(phages))]#line coloring values x axis\n",
    "    lys = [range(ow-lw+((bh+lw)*i),(ow+((bh+lw)*i))) for i in range(1,len(go))]#line coloring values y axis\n",
    "    \n",
    "    #white line dividers for different subspecies/clades\n",
    "    lysw = [range(ow-lw+((bh+lw)*(go.index(i)+1)),(ow+((bh+lw)*(go.index(i)+1)))) for i in gos]\n",
    "    \n",
    "    for x in lxs:\n",
    "        for xx in x:\n",
    "            for yy in range(height):\n",
    "                data[yy,xx]=color\n",
    "    for y in lys:\n",
    "        for yy in y:\n",
    "            for xx in range(width):\n",
    "                data[yy,xx]=color\n",
    "    for y in lysw:#dividers for clades\n",
    "        for yy in y:\n",
    "            for xx in range(width):\n",
    "                data[yy,xx]=black\n",
    "    ##color in outlines\n",
    "    for x in range(ow):\n",
    "        for y in range(height):\n",
    "            data[y,x]=color\n",
    "    for x in range(width-ow,width):\n",
    "        for y in range(height):\n",
    "            data[y,x]=color\n",
    "    for y in range(ow):\n",
    "        for x in range(width):\n",
    "            data[y,x]=color\n",
    "    for y in range(height-ow,height):\n",
    "        for x in range(width):\n",
    "            data[y,x]=color\n",
    "            \n",
    "    \n",
    "for genome in go:\n",
    "    if genome not in phd.keys():\n",
    "        gphc[genome] = 0\n",
    "        continue\n",
    "    for phage in phages:\n",
    "        if phage in phd[genome]:\n",
    "            pct = phd[genome][phage]\n",
    "            #data[1,10] = [255,0,0]red\n",
    "            #data[2,4] = [0,255,0]green\n",
    "            if pct<=(redpct+((100-redpct)/2)):\n",
    "                red = 255\n",
    "                green = round(((pct-redpct)/((100-redpct)/2))*(255))\n",
    "            elif pct > 50:\n",
    "                red = round(((100-pct)/((100-redpct)/2))*(255))\n",
    "                green = 255\n",
    "            color =[red, green, 0]\n",
    "            for h in range(ow+(go.index(genome)*(bh+lw)),ow+(go.index(genome)*(bh+lw))+bh):\n",
    "                for w in range(ow+(phages.index(phage)*(bw+lw)),ow+(phages.index(phage)*(bw+lw))+bw):\n",
    "                    data[h,w] = color\n",
    "            if phage not in phc.keys():\n",
    "                phc[phage] = 0\n",
    "            phc[phage] += 1\n",
    "            if genome not in gphc.keys():\n",
    "                gphc[genome] = 0\n",
    "            gphc[genome] += 1\n",
    "img = smp.toimage(data).save('2017-08-15_PHASTER_image_10PhageCutoff.png')\n",
    "img = smp.toimage(data)\n",
    "img.show()\n",
    "for i in phages:\n",
    "    print(phages.index(i)+1, i, ' '*(30-len(i)), 'count:', phc[i], '\\tproportion:', phc[i]/len(go))\n",
    "print('Min:', min([gphc[i] for i in gphc.keys()]), '\\tMax:', max([gphc[i] for i in gphc.keys()]),'\\tAvg:', \\\n",
    "      sum([gphc[i] for i in gphc.keys()])/len(go))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "320\n",
      "Average number of complete phages per genome: 1.3\n",
      "Total complete phages found: 577\n",
      "Total phages above 50% similarity to known phage in database: 577\n",
      "Number below cutoff: 0 \n",
      "Percent unknown: 0.0 %\n",
      "\n",
      "0 phages: 125\n",
      "1 phages: 157\n",
      "2 phages: 93\n",
      "3 phages: 55\n",
      "4 phages: 9\n",
      "5 phages: 3\n",
      "6 phages: 3\n",
      "\n",
      "Cutoff phage counts:\n",
      "Phage_Gifsy_1_NC_010392: 100\n",
      "Salmon_Fels_2_NC_010463: 90\n",
      "Salmon_SEN34_NC_028699: 50\n",
      "Phage_Gifsy_2_NC_010393: 47\n",
      "Salmon_118970_sal3_NC_031940: 37\n",
      "Entero_186_NC_001317: 23\n",
      "Salmon_SP_004_NC_021774: 21\n",
      "Entero_PsP3_NC_005340: 20\n",
      "Salmon_RE_2010_NC_019488: 19\n",
      "Haemop_HP2_NC_003315: 18\n",
      "Escher_D108_NC_013594: 13\n",
      "Entero_P88_NC_026014: 12\n",
      "Salmon_SEN1_NC_029003: 10\n",
      "Klebsi_phiKO2_NC_005857: 10\n",
      "Salmon_Fels_1_NC_010391: 9\n",
      "Salmon_epsilon15_NC_004775: 8\n",
      "Salmon_SJ46_NC_031129: 8\n",
      "Entero_P1_NC_005856: 7\n",
      "Shigel_SfII_NC_021857: 7\n",
      "Salmon_ST64B_NC_004313: 5\n",
      "Entero_fiAA91_ss_NC_022750: 5\n",
      "Escher_phiV10_NC_007804: 5\n",
      "Entero_Mu_NC_000929: 5\n",
      "Edward_GF_2_NC_026611: 4\n",
      "Salmon_SSU5_NC_018843: 4\n",
      "Entero_UAB_Phi20_NC_031019: 4\n",
      "Escher_pro483_NC_028943: 3\n",
      "Salmon_vB_SosS_Oslo_NC_018279: 3\n",
      "Salmon_SEN4_NC_029015: 3\n",
      "Vibrio_X29_NC_024369: 3\n",
      "Salmon_103203_sal5_NC_031946: 3\n",
      "Salmon_SPN1S_NC_016761: 2\n",
      "Entero_P22_NC_002371: 2\n",
      "Entero_Arya_NC_031048: 2\n",
      "Entero_SfV_NC_003444: 2\n",
      "Salmon_SEN5_NC_028701: 2\n",
      "Entero_P4_NC_001609: 2\n",
      "Entero_mEp460_NC_019716: 1\n",
      "Entero_Tyrion_NC_031077: 1\n",
      "Salmon_118970_sal4_NC_030919: 1\n",
      "Salmon_SPN3UB_NC_019545: 1\n",
      "Entero_ES18_NC_006949: 1\n",
      "Entero_lambda_NC_001416: 1\n",
      "Salmon_SP_031_NC_021775: 1\n",
      "Pseudo_MD8_NC_031091: 1\n",
      "Pasteu_F108_NC_008193: 1\n"
     ]
    }
   ],
   "source": [
    "#Phage Stats\n",
    "\n",
    "#strcutre of PHage Dictionary\n",
    "#phd[genome]={}\n",
    "#phd[genome][phage] = pct match\n",
    "\n",
    "#cutoff for phage calling\n",
    "cutoff = True\n",
    "co = 50\n",
    "\n",
    "\n",
    "##first, scrape the data\n",
    "phages = []\n",
    "phd = {} #phage dictionary\n",
    "phasterd = '/Users/jay.worley/Desktop/PHASTER/unzipped_results/'\n",
    "for i in [i for i in os.listdir(phasterd) if os.path.isdir(phasterd+i) and i in go]:\n",
    "    resultsf = phasterd+i+'/summary.txt'\n",
    "    with open(resultsf, 'r') as f:\n",
    "        rawresults = f.read()\n",
    "    results = rawresults.split('\\n\\n\\n\\n\\n')[1].strip().split('\\n')[3:]\n",
    "    \n",
    "    for result in results:\n",
    "        rl = ([i.strip() for i in result.split(' ') if len(i)>0])\n",
    "        if rl[2].startswith('intact'):\n",
    "            genome = rl[4].split('_')[0]\n",
    "            phage = '_'.join(rl[13].split('(')[0].split('_')[1:])\n",
    "            pct = float(rl[15].split('%')[0])\n",
    "            \n",
    "            #cutoff if called for\n",
    "            if cutoff == True and pct < co:\n",
    "                continue\n",
    "            phages.append(phage)\n",
    "            if genome not in phd.keys():\n",
    "                phd[genome] = {}\n",
    "            if phage not in phd[genome].keys():\n",
    "                phd[genome][phage] = pct\n",
    "            elif pct > phd[genome][phage]:\n",
    "                phd[genome][phage] = pct\n",
    "            else:\n",
    "                continue\n",
    "            continue\n",
    "phages = sorted(set(phages))\n",
    "\n",
    "#Average number of phages per genome:\n",
    "phage_counts = []\n",
    "num=genomes = 445\n",
    "print(len(phd))\n",
    "for key in phd.keys():\n",
    "    phage_counts.append(len(phd[key]))\n",
    "print('Average number of complete phages per genome:',round(sum(phage_counts)/445, 2))\n",
    "\n",
    "#total number of phages\n",
    "print('Total complete phages found:', sum(phage_counts))\n",
    "\n",
    "#nubmer of phages of each type scoring over 50% similarity\n",
    "\n",
    "cutoff_phages = []\n",
    "\n",
    "for key in phd.keys():\n",
    "    for phage in phd[key]:\n",
    "        if phd[key][phage] >= 50:\n",
    "            cutoff_phages.append(phage)\n",
    "print('Total phages above 50% similarity to known phage in database:', len(cutoff_phages))\n",
    "print('Number below cutoff:', sum(phage_counts) - len(cutoff_phages), '\\nPercent unknown:',\\\n",
    "     round(100*(sum(phage_counts)-len(cutoff_phages))/sum(phage_counts),1), '%\\n')\n",
    "\n",
    "#Breakdown numbers of genomes containing x number of complete phages\n",
    "phage_number = []\n",
    "for key in phd.keys():\n",
    "    phage_number.append(len(phd[key]))\n",
    "print('0 phages:',445-len(phage_number))\n",
    "for i in sorted(set(phage_number)):\n",
    "    print(i, 'phages:', phage_number.count(i))\n",
    "\n",
    "set_cutoff_phages = set(cutoff_phages)\n",
    "count_set_cutoff_phages = [(i, cutoff_phages.count(i)) for i in set_cutoff_phages]\n",
    "count_set_cutoff_phages = sorted(count_set_cutoff_phages, key = lambda x: x[1], reverse = True)\n",
    "print('\\nCutoff phage counts:')\n",
    "file_lines = []\n",
    "for hit in count_set_cutoff_phages:\n",
    "    print(hit[0]+':',hit[1])\n",
    "    file_lines.append(hit[0]+'\\t'+str(hit[1]))\n",
    "with open('2017-10-25_Phage_table.tab', 'w') as f:\n",
    "    f.write('\\n'.join(file_lines))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "##PLASMID FINDER OUTPUT##\n",
    "#Takes out individual subtypes - groups by incompatability group\n",
    "\n",
    "##first, scrape the data\n",
    "plasmids = []\n",
    "pld = {} #plasmid dictionary\n",
    "pff = './../LLNL-MvirDB/Plasmid_finder_hits_after_review.tab'#plasmid finder file\n",
    "with open(pff,'r') as f:\n",
    "    pft = f.read() #pftabs\n",
    "pfr = [i.strip() for i in pft.split('\\n')[1:] if len(i.strip())>0]#plasmid finder results\n",
    "for hit in pfr:\n",
    "    rl = hit.split('\\t')#result list\n",
    "    genome = rl[1].split('_')[0]\n",
    "    plasmid = rl[0].split('_')[0]\n",
    "    if '(' in plasmid:\n",
    "        pass\n",
    "    pct = float(rl[2]) #remember, pcts start at 80%\n",
    "    plasmids.append(plasmid)\n",
    "    if genome not in pld.keys():\n",
    "        pld[genome] = {}\n",
    "    if plasmid not in pld[genome].keys():\n",
    "        pld[genome][plasmid] = pct\n",
    "    elif pct > pld[genome][plasmid]:\n",
    "        pld[genome][plasmid] = pct\n",
    "    else:\n",
    "        continue\n",
    "    continue\n",
    "plasmids = sorted(set(plasmids))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 Col(Ye4449)      count: 1 \tproportion: 0.0022471910112359553\n",
      "2 Col156           count: 2 \tproportion: 0.0044943820224719105\n",
      "3 ColRNAI          count: 2 \tproportion: 0.0044943820224719105\n",
      "4 ColpVC           count: 6 \tproportion: 0.01348314606741573\n",
      "5 IncA/C           count: 4 \tproportion: 0.008988764044943821\n",
      "6 IncB/O/K/Z       count: 7 \tproportion: 0.015730337078651686\n",
      "7 IncFIB(K)        count: 1 \tproportion: 0.0022471910112359553\n",
      "8 IncFIB(pB171)    count: 17 \tproportion: 0.038202247191011236\n",
      "9 IncFIB(pLF82)    count: 2 \tproportion: 0.0044943820224719105\n",
      "10 IncFII           count: 28 \tproportion: 0.06292134831460675\n",
      "11 IncFII(29)       count: 56 \tproportion: 0.1258426966292135\n",
      "12 IncFII(Yp)       count: 3 \tproportion: 0.006741573033707865\n",
      "13 IncFII(p14)      count: 14 \tproportion: 0.03146067415730337\n",
      "14 IncFII(pCRY)     count: 6 \tproportion: 0.01348314606741573\n",
      "15 IncHI2           count: 1 \tproportion: 0.0022471910112359553\n",
      "16 IncHI2A          count: 1 \tproportion: 0.0022471910112359553\n",
      "17 IncI1            count: 14 \tproportion: 0.03146067415730337\n",
      "18 IncL/M(pMU407)   count: 4 \tproportion: 0.008988764044943821\n",
      "19 IncN             count: 1 \tproportion: 0.0022471910112359553\n",
      "20 IncQ1            count: 5 \tproportion: 0.011235955056179775\n",
      "21 IncX1            count: 7 \tproportion: 0.015730337078651686\n",
      "22 IncX2            count: 1 \tproportion: 0.0022471910112359553\n",
      "23 pSL483           count: 2 \tproportion: 0.0044943820224719105\n",
      "24 repA             count: 1 \tproportion: 0.0022471910112359553\n",
      "Min: 0 \tMax: 3 \tAvg: 0.41797752808988764\n",
      "\n",
      "0 plasmids: 289\n",
      "1 plasmids 130\n",
      "2 plasmids 21\n",
      "3 plasmids 5\n"
     ]
    }
   ],
   "source": [
    "##NOW CREATE THE IMAGE##\n",
    "#This includes all plasmids\n",
    "\n",
    "plc = {} #plasmid counts\n",
    "gplc = {}\n",
    "#create a matrix of the image size\n",
    "#((row, col), pct) -> ((genome,plasmid),match)\n",
    "\n",
    "##pixel dimensions\n",
    "bh = 10 #box height\n",
    "bw = 40 #box width\n",
    "lw = 3 #line width\n",
    "ow = 5 #outline width\n",
    "black = [0,0,0]\n",
    "white = [255,255,255]\n",
    "grey = [100,100,100]\n",
    "\n",
    "height = (len(go)*(bh+lw))-lw+(2*ow)\n",
    "width = (len(plasmids)*(bw+lw))-lw+(2*ow)\n",
    "\n",
    "data = np.zeros((height,width,3), np.uint8)\n",
    "\n",
    "##color lines white\n",
    "\n",
    "color_my_lines = True\n",
    "if color_my_lines == True:\n",
    "    color = black\n",
    "    lxs = [range(ow-lw+((bw+lw)*i),(ow+((bw+lw)*i))) for i in range(1,len(plasmids))]#line coloring values x axis\n",
    "    lys = [range(ow-lw+((bh+lw)*i),(ow+((bh+lw)*i))) for i in range(1,len(go))]#line coloring values y axis\n",
    "    \n",
    "    #white line dividers for different subspecies/clades\n",
    "    lysw = [range(ow-lw+((bh+lw)*(go.index(i)+1)),(ow+((bh+lw)*(go.index(i)+1)))) for i in gos]\n",
    "    \n",
    "    for x in lxs:\n",
    "        for xx in x:\n",
    "            for yy in range(height):\n",
    "                data[yy,xx]=color\n",
    "    for y in lys:\n",
    "        for yy in y:\n",
    "            for xx in range(width):\n",
    "                data[yy,xx]=color\n",
    "    for y in lysw:#dividers for clades\n",
    "        for yy in y:\n",
    "            for xx in range(width):\n",
    "                data[yy,xx]=color            \n",
    "    \n",
    "    ##color in outlines\n",
    "    for x in range(ow):\n",
    "        for y in range(height):\n",
    "            data[y,x]=color\n",
    "    for x in range(width-ow,width):\n",
    "        for y in range(height):\n",
    "            data[y,x]=color\n",
    "    for y in range(ow):\n",
    "        for x in range(width):\n",
    "            data[y,x]=color\n",
    "    for y in range(height-ow,height):\n",
    "        for x in range(width):\n",
    "            data[y,x]=color\n",
    "\n",
    "for genome in go:\n",
    "    if genome not in pld.keys():\n",
    "        gplc[genome] = 0\n",
    "        continue\n",
    "    for plasmid in plasmids:\n",
    "        if plasmid in pld[genome]:\n",
    "            pct = pld[genome][plasmid]\n",
    "            #using yellow to green since percentage hits start at 80, so we have some sorting already by score\n",
    "            #data[1,10] = [255,0,0]red\n",
    "            #data[2,4] = [0,255,0]green\n",
    "            red = round(255*((100-pct)/20))\n",
    "            green = 255\n",
    "            color =[red, green, 0]\n",
    "            for h in range(ow+(go.index(genome)*(bh+lw)),ow+(go.index(genome)*(bh+lw))+bh):\n",
    "                for w in range(ow+(plasmids.index(plasmid)*(bw+lw)),ow+(plasmids.index(plasmid)*(bw+lw))+bw):\n",
    "                    data[h,w] = color\n",
    "            if plasmid not in plc.keys():\n",
    "                plc[plasmid] = 0\n",
    "            plc[plasmid] += 1\n",
    "            if genome not in gplc.keys():\n",
    "                gplc[genome] = 0\n",
    "            gplc[genome] += 1\n",
    "img = smp.toimage(data).save('2018-08-06_PLASMIDFINDER_image_un_compressed.png')\n",
    "img = smp.toimage(data)\n",
    "img.show()\n",
    "for i in plasmids:\n",
    "    print(plasmids.index(i)+1, i, ' '*(15-len(i)), 'count:', plc[i], '\\tproportion:', plc[i]/len(go))\n",
    "print('Min:', min([gplc[i] for i in gplc.keys()]), '\\tMax:', max([gplc[i] for i in gplc.keys()]),'\\tAvg:', \\\n",
    "      sum([gplc[i] for i in gplc.keys()])/len(go))\n",
    "\n",
    "#Breakdown numbers of genomes containing x number of complete phages\n",
    "plasmid_number = []\n",
    "for key in pld.keys():\n",
    "    plasmid_number.append(len(pld[key]))\n",
    "print('\\n0 plasmids:',445-len(plasmid_number))\n",
    "for i in sorted(set(plasmid_number)):\n",
    "    print(i,'plasmids',plasmid_number.count(i))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 IncFIB(K)        count: 1 \tproportion: 0.0022471910112359553\n",
      "2 IncFIB(pB171)    count: 18 \tproportion: 0.04044943820224719\n",
      "3 IncFIB(pLF82)    count: 2 \tproportion: 0.0044943820224719105\n",
      "4 IncFII           count: 28 \tproportion: 0.06292134831460675\n",
      "5 IncFII(29)       count: 56 \tproportion: 0.1258426966292135\n",
      "6 IncFII(Yp)       count: 3 \tproportion: 0.006741573033707865\n",
      "7 IncFII(p14)      count: 14 \tproportion: 0.03146067415730337\n",
      "8 IncFII(pCRY)     count: 7 \tproportion: 0.015730337078651686\n",
      "Min: 0 \tMax: 2 \tAvg: 0.2898876404494382\n",
      "\n",
      "0 plasmids: 328\n",
      "1 plasmids 104\n",
      "2 plasmids 13\n"
     ]
    }
   ],
   "source": [
    "##NOW CREATE THE IMAGE##\n",
    "#This has ONLY INC F MEMBERS\n",
    "\n",
    "pldF = {} #plasmid dictionary with only incF or incFII members\n",
    "Fplasmids = [] #plasmids in incF or incFII\n",
    "\n",
    "for genome in pld.keys():\n",
    "    incF_plasmids = [i for i in pld[genome] if i.startswith('IncF')]\n",
    "    if len(incF_plasmids) < 1:\n",
    "        continue #pass by those with no incF or incFII plasmids\n",
    "    if genome not in pldF:\n",
    "        pldF[genome] = {}\n",
    "    for plasmid in [i for i in pld[genome] if i.startswith('IncF')]:\n",
    "        Fplasmids.append(plasmid)\n",
    "        pldF[genome][plasmid] = pld[genome][plasmid]\n",
    "Fplasmids = list(sorted(set(Fplasmids))) #to get the right height\n",
    "\n",
    "plc = {} #plasmid counts\n",
    "gplc = {}\n",
    "#create a matrix of the image size\n",
    "#((row, col), pct) -> ((genome,plasmid),match)\n",
    "\n",
    "##pixel dimensions\n",
    "bh = 10 #box height\n",
    "bw = 40 #box width\n",
    "lw = 0 #line width\n",
    "ow = 0 #outline width\n",
    "black = [0,0,0]\n",
    "white = [255,255,255]\n",
    "grey = [100,100,100]\n",
    "\n",
    "height = (len(go)*(bh+lw))-lw+(2*ow)\n",
    "width = (len(Fplasmids)*(bw+lw))-lw+(2*ow)\n",
    "\n",
    "data = np.zeros((height,width,3), np.uint8)\n",
    "\n",
    "##color lines white\n",
    "\n",
    "color_my_lines = True\n",
    "if color_my_lines == True:\n",
    "    color = grey\n",
    "    lxs = [range(ow-lw+((bw+lw)*i),(ow+((bw+lw)*i))) for i in range(1,len(Fplasmids))]#line coloring values x axis\n",
    "    lys = [range(ow-lw+((bh+lw)*i),(ow+((bh+lw)*i))) for i in range(1,len(go))]#line coloring values y axis\n",
    "    \n",
    "    #white line dividers for different subspecies/clades\n",
    "    lysw = [range(ow-lw+((bh+lw)*(go.index(i)+1)),(ow+((bh+lw)*(go.index(i)+1)))) for i in gos]\n",
    "    \n",
    "    for x in lxs:\n",
    "        for xx in x:\n",
    "            for yy in range(height):\n",
    "                data[yy,xx]=color\n",
    "    for y in lys:\n",
    "        for yy in y:\n",
    "            for xx in range(width):\n",
    "                data[yy,xx]=color\n",
    "    for y in lysw:#dividers for clades\n",
    "        for yy in y:\n",
    "            for xx in range(width):\n",
    "                data[yy,xx]=white            \n",
    "    \n",
    "    ##color in outlines\n",
    "    for x in range(ow):\n",
    "        for y in range(height):\n",
    "            data[y,x]=color\n",
    "    for x in range(width-ow,width):\n",
    "        for y in range(height):\n",
    "            data[y,x]=color\n",
    "    for y in range(ow):\n",
    "        for x in range(width):\n",
    "            data[y,x]=color\n",
    "    for y in range(height-ow,height):\n",
    "        for x in range(width):\n",
    "            data[y,x]=color\n",
    "\n",
    "for genome in go:\n",
    "    if genome not in pldF.keys():\n",
    "        gplc[genome] = 0\n",
    "        continue\n",
    "    for plasmid in Fplasmids:\n",
    "        if plasmid in pldF[genome]:\n",
    "            pct = pldF[genome][plasmid]\n",
    "            #using yellow to green since percentage hits start at 80, so we have some sorting already by score\n",
    "            #data[1,10] = [255,0,0]red\n",
    "            #data[2,4] = [0,255,0]green\n",
    "            red = round(255*((100-pct)/20))\n",
    "            green = 255\n",
    "            color =[red, green, 0]\n",
    "            for h in range(ow+(go.index(genome)*(bh+lw)),ow+(go.index(genome)*(bh+lw))+bh):\n",
    "                for w in range(ow+(Fplasmids.index(plasmid)*(bw+lw)),ow+(Fplasmids.index(plasmid)*(bw+lw))+bw):\n",
    "                    data[h,w] = color\n",
    "            if plasmid not in plc.keys():\n",
    "                plc[plasmid] = 0\n",
    "            plc[plasmid] += 1\n",
    "            if genome not in gplc.keys():\n",
    "                gplc[genome] = 0\n",
    "            gplc[genome] += 1\n",
    "img = smp.toimage(data).save('2017-08-23_PLASMIDFINDER_IncF_and_FII_Only.png')\n",
    "img = smp.toimage(data)\n",
    "img.show()\n",
    "for i in Fplasmids:\n",
    "    print(Fplasmids.index(i)+1, i, ' '*(15-len(i)), 'count:', plc[i], '\\tproportion:', plc[i]/len(go))\n",
    "print('Min:', min([gplc[i] for i in gplc.keys()]), '\\tMax:', max([gplc[i] for i in gplc.keys()]),'\\tAvg:', \\\n",
    "      sum([gplc[i] for i in gplc.keys()])/len(go))\n",
    "\n",
    "#Breakdown numbers of genomes containing x number of complete phages\n",
    "plasmid_number = []\n",
    "for key in pldF.keys():\n",
    "    plasmid_number.append(len(pldF[key]))\n",
    "print('\\n0 plasmids:',445-len(plasmid_number))\n",
    "for i in sorted(set(plasmid_number)):\n",
    "    print(i,'plasmids',plasmid_number.count(i))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['IncFII(pCRY)', 'IncFIB(K)', 'IncFII(Yp)', 'IncFII(29)', 'IncFIB(pLF82)', 'IncFII(p14)', 'IncFII', 'IncFIB(pB171)']\n",
      "{'SEQ000031518': {'IncFII(pCRY)': 95.262, 'IncFII': 82.625}, 'SEQ000009073': {'IncFII': 82.996}, 'SEQ000041121': {'IncFII(29)': 81.985}, 'SEQ000014948': {'IncFII': 88.213}, 'SEQ000014048': {'IncFII(29)': 83.206}, 'SEQ000007942': {'IncFII(Yp)': 80.973}, 'SEQ000029384': {'IncFII(29)': 83.206, 'IncFII': 81.203}, 'SEQ000018161': {'IncFII(p14)': 87.55}, 'SEQ000014053': {'IncFII(29)': 83.206}, 'SEQ000004780': {'IncFII(29)': 83.206}, 'SEQ000034189': {'IncFII(29)': 83.206}, 'SEQ000015178': {'IncFIB(pB171)': 85.047}, 'SEQ000031933': {'IncFII': 82.996}, 'SEQ000036900': {'IncFII(29)': 83.206, 'IncFIB(pB171)': 90.669}, 'SEQ000015331': {'IncFII(29)': 83.784}, 'SEQ000015070': {'IncFII(p14)': 93.927}, 'SEQ000015057': {'IncFII(29)': 84.556}, 'SEQ000008898': {'IncFII(29)': 83.784}, 'SEQ000007959': {'IncFII(29)': 84.231}, 'SEQ000005604': {'IncFIB(pB171)': 92.702}, 'SEQ000003858': {'IncFIB(pB171)': 92.224}, 'SEQ000011765': {'IncFII(29)': 84.211}, 'SEQ000030783': {'IncFIB(pB171)': 85.514}, 'SAMN02844917': {'IncFII(pCRY)': 83.833, 'IncFII(29)': 83.206}, 'SEQ000015055': {'IncFII': 82.996}, 'SEQ000037531': {'IncFII(29)': 84.17}, 'SEQ000005318': {'IncFII(29)': 83.206}, 'SEQ000034332': {'IncFII(p14)': 91.188}, 'SEQ000026078': {'IncFII(Yp)': 80.973}, 'SEQ000008928': {'IncFII': 93.462}, 'SEQ000034547': {'IncFII(p14)': 91.188}, 'SEQ000005639': {'IncFIB(pB171)': 90.98}, 'SEQ000034510': {'IncFII(29)': 84.556}, 'SEQ000006276': {'IncFII(29)': 84.17}, 'SEQ000015056': {'IncFII(29)': 84.556}, 'SEQ000013953': {'IncFII(29)': 83.784}, 'SEQ000034319': {'IncFII(p14)': 91.188}, 'SEQ000012956': {'IncFII': 82.996}, 'SEQ000011867': {'IncFII(p14)': 83.544}, 'SEQ000036044': {'IncFII': 86.328, 'IncFIB(pB171)': 90.98}, 'SEQ000038403': {'IncFII': 82.996}, 'SEQ000014122': {'IncFII(p14)': 81.25}, 'SEQ000014119': {'IncFII(29)': 84.231}, 'SEQ000038372': {'IncFII(pCRY)': 87.0}, 'SEQ000029488': {'IncFII': 82.996}, 'SEQ000034752': {'IncFII(29)': 83.206}, 'SEQ000034548': {'IncFII(p14)': 91.188}, 'SEQ000038406': {'IncFII': 82.996}, 'SEQ000031524': {'IncFII': 84.496}, 'SEQ000027790': {'IncFII(29)': 83.784}, 'SEQ000011949': {'IncFII(Yp)': 80.973}, 'SEQ000028559': {'IncFII(29)': 83.784}, 'SEQ000011998': {'IncFII(pCRY)': 98.387, 'IncFII(p14)': 92.218}, 'SEQ000009069': {'IncFII': 82.996}, 'SEQ000015496': {'IncFII': 82.996}, 'SEQ000036074': {'IncFII': 86.328}, 'SEQ000039242': {'IncFIB(pLF82)': 87.004}, 'SEQ000015887': {'IncFII': 84.052}, 'SEQ000036053': {'IncFIB(K)': 98.929}, 'SEQ000005500': {'IncFIB(pB171)': 92.068}, 'SEQ000038395': {'IncFII(29)': 83.27}, 'SEQ000014645': {'IncFIB(pB171)': 92.224}, 'SEQ000015395': {'IncFII(pCRY)': 95.11}, 'SEQ000000407': {'IncFIB(pB171)': 93.157}, 'SEQ000014862': {'IncFII(29)': 83.206}, 'SEQ000011827': {'IncFII(29)': 84.17}, 'SEQ000028561': {'IncFIB(pB171)': 92.702}, 'SEQ000033581': {'IncFII(29)': 84.17}, 'SEQ000027650': {'IncFIB(pB171)': 90.98}, 'SEQ000000272': {'IncFII(29)': 83.206}, 'SEQ000030732': {'IncFII': 82.996}, 'SEQ000013962': {'IncFII(29)': 82.824}, 'SEQ000005035': {'IncFII(29)': 83.206, 'IncFIB(pB171)': 85.67}, 'SEQ000004791': {'IncFII(29)': 83.206}, 'SEQ000000373': {'IncFII(29)': 84.17}, 'SEQ000011950': {'IncFII(Yp)': 80.973}, 'SEQ000015445': {'IncFII(p14)': 80.695}, 'SEQ000034996': {'IncFII(pCRY)': 94.266}, 'SEQ000014563': {'IncFII': 82.625}, 'SEQ000012954': {'IncFII(29)': 83.398}, 'SEQ000017189': {'IncFII(29)': 84.17, 'IncFIB(pB171)': 97.183}, 'SEQ000006821': {'IncFII(29)': 84.17}, 'SEQ000004769': {'IncFII(29)': 83.27}, 'SEQ000012291': {'IncFII(29)': 82.824}, 'SEQ000006979': {'IncFII': 82.996}, 'SEQ000004956': {'IncFII(29)': 83.588}, 'SEQ000028558': {'IncFII(29)': 83.784}, 'SEQ000027234': {'IncFII(29)': 83.206}, 'SEQ000003055': {'IncFII(29)': 83.784}, 'SEQ000008884': {'IncFII(29)': 80.97, 'IncFII': 83.401}, 'SEQ000014151': {'IncFII': 82.625}, 'SEQ000033571': {'IncFII(p14)': 80.545}, 'SEQ000014520': {'IncFII(29)': 84.17}, 'SEQ000004718': {'IncFII': 83.398}, 'SEQ000003035': {'IncFIB(pLF82)': 86.823}, 'SEQ000037533': {'IncFII(29)': 83.206}, 'SEQ000022441': {'IncFII': 86.328}, 'SEQ000015955': {'IncFII': 84.052}, 'SEQ000012073': {'IncFII(29)': 83.206}, 'SEQ000014861': {'IncFII(29)': 83.206}, 'SEQ000014510': {'IncFII(29)': 82.824}, 'SEQ000014785': {'IncFII(pCRY)': 94.941}, 'SEQ000037525': {'IncFII(29)': 81.985}, 'SEQ000017775': {'IncFII(p14)': 90.076}, 'SEQ000005315': {'IncFII(29)': 83.206}, 'SEQ000010891': {'IncFII(p14)': 80.545}, 'SEQ000031188': {'IncFII(29)': 82.89, 'IncFII': 82.996}, 'SEQ000013945': {'IncFII(29)': 82.824}, 'SEQ000014042': {'IncFII(p14)': 92.218, 'IncFIB(pB171)': 85.514}, 'SEQ000014500': {'IncFII(29)': 82.824}, 'SEQ000013908': {'IncFII(29)': 84.17}, 'SEQ000014373': {'IncFII(29)': 84.17, 'IncFIB(pB171)': 90.047}, 'SEQ000036178': {'IncFIB(pB171)': 91.757}, 'SEQ000015506': {'IncFII(29)': 83.206}, 'SEQ000011807': {'IncFII(29)': 84.211, 'IncFIB(pB171)': 94.712}, 'SEQ000015186': {'IncFII': 85.938}, 'SAMN02844600': {'IncFII': 80.816}}\n"
     ]
    }
   ],
   "source": [
    "pldF = {} #plasmid dictionary with only incF or incFII members\n",
    "Fplasmids = [] #plasmids in incF or incFII\n",
    "\n",
    "for genome in pld.keys():\n",
    "    incF_plasmids = [i for i in pld[genome] if i.startswith('IncF')]\n",
    "    if len(incF_plasmids) < 1:\n",
    "        continue #pass by those with no incF or incFII plasmids\n",
    "    if genome not in pldF:\n",
    "        pldF[genome] = {}\n",
    "    for plasmid in [i for i in pld[genome] if i.startswith('IncF')]:\n",
    "        Fplasmids.append(plasmid)\n",
    "        pldF[genome][plasmid] = pld[genome][plasmid]\n",
    "Fplasmids = list(set(Fplasmids))\n",
    "print(Fplasmids)\n",
    "print(pldF)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "##VFDB BLAST HITS##\n",
    "\n",
    "#For this, we pull from another analysis that eliminated duplicates\n",
    "\n",
    "wb = op.load_workbook('/Users/jay.worley/Desktop/LLNL-MvirDB/testbook.xlsx')\n",
    "ws = wb['Sheet']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "##pixel dimensions\n",
    "bh = 10 #box height\n",
    "bw = 40 #box width\n",
    "lw = 0 #line width\n",
    "ow = 0 #outline width\n",
    "black = [0,0,0]\n",
    "white = [255,255,255]\n",
    "grey = [100,100,100]\n",
    "\n",
    "#vg class indices in excel file\n",
    "class_inds = [i+1 for i in range(ws.max_column) if ws.cell(column = i+1, row = 1).value is not None]\n",
    "class_inds.append(ws.max_column+2)\n",
    "\n",
    "color_my_lines = True\n",
    "\n",
    "vgc = {}\n",
    "gvgc =  {}\n",
    "for ind in class_inds[:-1]:\n",
    "    class_name = ws.cell(column=ind, row=1).value\n",
    "    outfile = '2017-05-17_VFDB_'+class_name+'_image.png'\n",
    "    start = ind\n",
    "    stop = class_inds[class_inds.index(ind)+1]\n",
    "    \n",
    "    height = ((ws.max_row-3)*(bh+lw))-lw+(2*ow)\n",
    "    width = ((stop-start)*(bw+lw))-lw+(2*ow)\n",
    "\n",
    "    data = np.zeros((height,width,3), np.uint8)\n",
    "\n",
    "    if color_my_lines == True:\n",
    "        color = grey\n",
    "        lxs = [range(ow-lw+((bw+lw)*i),(ow+((bw+lw)*i))) for i in range(1,stop-start)]#line coloring values x axis\n",
    "        lys = [range(ow-lw+((bh+lw)*i),(ow+((bh+lw)*i))) for i in range(1,ws.max_row-3)]#line coloring values y axis\n",
    "        \n",
    "        #white line dividers for different subspecies/clades\n",
    "        lysw = [range(ow-lw+((bh+lw)*(go.index(i)+1)),(ow+((bh+lw)*(go.index(i)+1)))) for i in gos]\n",
    "        \n",
    "        for x in lxs:\n",
    "            for xx in x:\n",
    "                for yy in range(height):\n",
    "                    data[yy,xx]=color\n",
    "        for y in lys:\n",
    "            for yy in y:\n",
    "                for xx in range(width):\n",
    "                    data[yy,xx]=color\n",
    "        for y in lysw:#Dividers for clades\n",
    "            for yy in y:\n",
    "                for xx in range(width):\n",
    "                    data[yy,xx]=white \n",
    "                    \n",
    "        ##color in outlines\n",
    "        for x in range(ow):\n",
    "            for y in range(height):\n",
    "                data[y,x]=color\n",
    "        for x in range(width-ow,width):\n",
    "            for y in range(height):\n",
    "                data[y,x]=color\n",
    "        for y in range(ow):\n",
    "            for x in range(width):\n",
    "                data[y,x]=color\n",
    "        for y in range(height-ow,height):\n",
    "            for x in range(width):\n",
    "                data[y,x]=color\n",
    "    \n",
    "    for i in range(stop-start):\n",
    "        gene = ws.cell(column=i+start,row=3).value\n",
    "        geneclass = ws.cell(column=i+start, row=2).value\n",
    "        vgc[gene] = 0\n",
    "        \n",
    "        ##if new class and not the first, color in lines to the left\n",
    "        if (i != 0) and (geneclass != None):\n",
    "            for xx in range(ow-lw+((bw+lw)*i),(ow+((bw+lw)*i))):\n",
    "                for yy in range(height):\n",
    "                    data[yy,xx]=white\n",
    "            \n",
    "        \n",
    "        for j in range(ws.max_row-3):\n",
    "            genome = ws.cell(column=1,row=j+4).value\n",
    "            if genome not in gvgc.keys():\n",
    "                gvgc[genome] = 0\n",
    "            if ws.cell(column=i+start,row=j+4).value is not None:\n",
    "                pct = float(ws.cell(column=i+start,row=j+4).value)\n",
    "                red = round(255*((100-pct)/20))\n",
    "                if red > 255:\n",
    "                    red == 255\n",
    "                green = 255\n",
    "                color =[red, green, 0]\n",
    "                for h in range(ow+(go.index(genome)*(bh+lw)),ow+(go.index(genome)*(bh+lw))+bh):\n",
    "                    for w in range(ow+(i*(bw+lw)),ow+(i*(bw+lw))+bw):\n",
    "                        data[h,w] = color\n",
    "                vgc[gene] += 1\n",
    "                gvgc[genome] += 1\n",
    "    img = smp.toimage(data).save(outfile)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[(37, 42), (58, 62), (26, 30), (82, 89), (33, 37), (96, 102), (111, 116), (62, 69), (18, 22)]\n",
      "3577 1995 46\n"
     ]
    }
   ],
   "source": [
    "#Specific to figure of specifically compared adherence genes\n",
    "\n",
    "##pixel dimensions\n",
    "bh = 5 #box height\n",
    "bw = 40 #box width\n",
    "lw = 3 #line width\n",
    "ow = 10 #outline width\n",
    "black = [0,0,0]\n",
    "white = [255,255,255]\n",
    "grey = [100,100,100]\n",
    "\n",
    "select_genes = sorted(['Lpf','Peg','Saf','Sta','Stc','Ste','Stj','Stk','Tcf'])\n",
    "\n",
    "class_inds = []\n",
    "class_starts = []\n",
    "#vg class indices in excel file\n",
    "for gene in select_genes:\n",
    "    for i in range (ws.max_column):\n",
    "        if ws.cell(column = i+2, row = 2).value == gene:\n",
    "            class_starts.append(i+2)\n",
    "            break\n",
    "for i in class_starts:\n",
    "    br = 0\n",
    "    j = i\n",
    "    while br == 0:\n",
    "        j += 1\n",
    "        if ws.cell(column=j+1, row=2).value is not None:\n",
    "            class_inds.append((i,j+1))\n",
    "            br = 1\n",
    "            continue\n",
    "        continue\n",
    "color_my_lines = True\n",
    "\n",
    "vgc = {}\n",
    "gvgc =  {}\n",
    "print(class_inds)\n",
    "\n",
    "#DON\"T KICK OFF LOOP - ONE LOOP, ONE IMAGE!\n",
    "class_name = 'Specific_adherence_genes_independent_evo'\n",
    "outfile = '2017-05-17_VFDB_'+class_name+'_image.png'\n",
    "total_cols = sum([(tup[1]-tup[0]) for tup in class_inds])\n",
    "    \n",
    "height = ((ws.max_row-3)*(bh+lw))-lw+(2*ow)\n",
    "width = ((total_cols)*(bw+lw))-lw+(2*ow)\n",
    "print(height, width, total_cols)\n",
    "\n",
    "data = np.zeros((height,width,3), np.uint8)\n",
    "\n",
    "if color_my_lines == True:\n",
    "    color = grey\n",
    "    lxs = [range(ow-lw+((bw+lw)*i),(ow+((bw+lw)*i))) for i in range(1,total_cols)]#line coloring values x axis\n",
    "    lys = [range(ow-lw+((bh+lw)*i),(ow+((bh+lw)*i))) for i in range(1,ws.max_row-3)]#line coloring values y axis\n",
    "        \n",
    "    #white line dividers for different subspecies/clades\n",
    "    lysw = [range(ow-lw+((bh+lw)*(go.index(i)+1)),(ow+((bh+lw)*(go.index(i)+1)))) for i in gos]\n",
    "        \n",
    "    for x in lxs:\n",
    "        for xx in x:\n",
    "            for yy in range(height):\n",
    "                data[yy,xx]=color\n",
    "    for y in lys:\n",
    "        for yy in y:\n",
    "            for xx in range(width):\n",
    "                data[yy,xx]=color\n",
    "    for y in lysw:#Dividers for clades\n",
    "        for yy in y:\n",
    "            for xx in range(width):\n",
    "                data[yy,xx]=white \n",
    "                    \n",
    "    ##color in outlines\n",
    "    for x in range(ow):\n",
    "        for y in range(height):\n",
    "            data[y,x]=color\n",
    "    for x in range(width-ow,width):\n",
    "        for y in range(height):\n",
    "            data[y,x]=color\n",
    "    for y in range(ow):\n",
    "        for x in range(width):\n",
    "            data[y,x]=color\n",
    "    for y in range(height-ow,height):\n",
    "        for x in range(width):\n",
    "            data[y,x]=color\n",
    "\n",
    "gene_inds = []\n",
    "for i in class_inds:\n",
    "    gene_inds = gene_inds+list(range(i[0],i[1]))\n",
    "    \n",
    "column_num = 0\n",
    "for i in gene_inds:\n",
    "    gene = ws.cell(column=i,row=3).value\n",
    "    geneclass = ws.cell(column=i, row=2).value\n",
    "    vgc[gene] = 0\n",
    "        \n",
    "    ##if new class and not the first, color in lines to the left\n",
    "    if (column_num != 0) and (geneclass != None):\n",
    "        for xx in range(ow-lw+((bw+lw)*column_num),(ow+((bw+lw)*column_num))):\n",
    "            for yy in range(height):\n",
    "                data[yy,xx]=white\n",
    "            \n",
    "        \n",
    "    for j in range(ws.max_row-3):\n",
    "        genome = ws.cell(column=1,row=j+4).value\n",
    "        if genome not in gvgc.keys():\n",
    "            gvgc[genome] = 0\n",
    "        if ws.cell(column=i,row=j+4).value is not None:\n",
    "            pct = float(ws.cell(column=i,row=j+4).value)\n",
    "            red = round(255*((100-pct)/20))\n",
    "            if red > 255:\n",
    "                red == 255\n",
    "            green = 255\n",
    "            color =[red, green, 0]\n",
    "            for h in range(ow+(go.index(genome)*(bh+lw)),ow+(go.index(genome)*(bh+lw))+bh):\n",
    "                for w in range(ow+(column_num*(bw+lw)),ow+(column_num*(bw+lw))+bw):\n",
    "                    data[h,w] = color\n",
    "            vgc[gene] += 1\n",
    "            gvgc[genome] += 1\n",
    "    column_num += 1\n",
    "img = smp.toimage(data).save(outfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for ind in class_inds[:-1]:\n",
    "    class_name = ws.cell(column=ind, row=1).value\n",
    "    outfile = '2017-05-17_VFDB_'+class_name+'_image.png'\n",
    "    start = ind\n",
    "    stop = class_inds[class_inds.index(ind)+1]\n",
    "    \n",
    "    height = ((ws.max_row-3)*(bh+lw))-lw+(2*ow)\n",
    "    width = ((stop-start)*(bw+lw))-lw+(2*ow)\n",
    "\n",
    "    data = np.zeros((height,width,3), np.uint8)\n",
    "\n",
    "    if color_my_lines == True:\n",
    "        color = grey\n",
    "        lxs = [range(ow-lw+((bw+lw)*i),(ow+((bw+lw)*i))) for i in range(1,stop-start)]#line coloring values x axis\n",
    "        lys = [range(ow-lw+((bh+lw)*i),(ow+((bh+lw)*i))) for i in range(1,ws.max_row-3)]#line coloring values y axis\n",
    "        \n",
    "        #white line dividers for different subspecies/clades\n",
    "        lysw = [range(ow-lw+((bh+lw)*(go.index(i)+1)),(ow+((bh+lw)*(go.index(i)+1)))) for i in gos]\n",
    "        \n",
    "        for x in lxs:\n",
    "            for xx in x:\n",
    "                for yy in range(height):\n",
    "                    data[yy,xx]=color\n",
    "        for y in lys:\n",
    "            for yy in y:\n",
    "                for xx in range(width):\n",
    "                    data[yy,xx]=color\n",
    "        for y in lysw:#Dividers for clades\n",
    "            for yy in y:\n",
    "                for xx in range(width):\n",
    "                    data[yy,xx]=white \n",
    "                    \n",
    "        ##color in outlines\n",
    "        for x in range(ow):\n",
    "            for y in range(height):\n",
    "                data[y,x]=color\n",
    "        for x in range(width-ow,width):\n",
    "            for y in range(height):\n",
    "                data[y,x]=color\n",
    "        for y in range(ow):\n",
    "            for x in range(width):\n",
    "                data[y,x]=color\n",
    "        for y in range(height-ow,height):\n",
    "            for x in range(width):\n",
    "                data[y,x]=color\n",
    "    \n",
    "    for i in range(stop-start):\n",
    "        gene = ws.cell(column=i+start,row=3).value\n",
    "        geneclass = ws.cell(column=i+start, row=2).value\n",
    "        vgc[gene] = 0\n",
    "        \n",
    "        ##if new class and not the first, color in lines to the left\n",
    "        if (i != 0) and (geneclass != None):\n",
    "            for xx in range(ow-lw+((bw+lw)*i),(ow+((bw+lw)*i))):\n",
    "                for yy in range(height):\n",
    "                    data[yy,xx]=white\n",
    "            \n",
    "        \n",
    "        for j in range(ws.max_row-3):\n",
    "            genome = ws.cell(column=1,row=j+4).value\n",
    "            if genome not in gvgc.keys():\n",
    "                gvgc[genome] = 0\n",
    "            if ws.cell(column=i+start,row=j+4).value is not None:\n",
    "                pct = float(ws.cell(column=i+start,row=j+4).value)\n",
    "                red = round(255*((100-pct)/20))\n",
    "                if red > 255:\n",
    "                    red == 255\n",
    "                green = 255\n",
    "                color =[red, green, 0]\n",
    "                for h in range(ow+(go.index(genome)*(bh+lw)),ow+(go.index(genome)*(bh+lw))+bh):\n",
    "                    for w in range(ow+(i*(bw+lw)),ow+(i*(bw+lw))+bw):\n",
    "                        data[h,w] = color\n",
    "                vgc[gene] += 1\n",
    "                gvgc[genome] += 1\n",
    "    img = smp.toimage(data).save(outfile)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 -                count: 0 \tproportion: 0.0\n",
      "2 avrA             count: 343 \tproportion: 0.7707865168539326\n",
      "3 bcfA             count: 223 \tproportion: 0.501123595505618\n",
      "4 bcfB             count: 216 \tproportion: 0.4853932584269663\n",
      "5 bcfC             count: 437 \tproportion: 0.9820224719101124\n",
      "6 bcfD             count: 437 \tproportion: 0.9820224719101124\n",
      "7 bcfE             count: 437 \tproportion: 0.9820224719101124\n",
      "8 bcfF             count: 217 \tproportion: 0.48764044943820223\n",
      "9 bcfG             count: 436 \tproportion: 0.9797752808988764\n",
      "10 cdtB             count: 176 \tproportion: 0.3955056179775281\n",
      "11 csgA             count: 445 \tproportion: 1.0\n",
      "12 csgB             count: 445 \tproportion: 1.0\n",
      "13 csgC             count: 445 \tproportion: 1.0\n",
      "14 csgD             count: 445 \tproportion: 1.0\n",
      "15 csgE             count: 444 \tproportion: 0.9977528089887641\n",
      "16 csgF             count: 445 \tproportion: 1.0\n",
      "17 csgG             count: 445 \tproportion: 1.0\n",
      "18 fimA             count: 444 \tproportion: 0.9977528089887641\n",
      "19 fimC             count: 444 \tproportion: 0.9977528089887641\n",
      "20 fimD             count: 444 \tproportion: 0.9977528089887641\n",
      "21 fimF             count: 443 \tproportion: 0.9955056179775281\n",
      "22 fimH             count: 444 \tproportion: 0.9977528089887641\n",
      "23 fimI             count: 442 \tproportion: 0.9932584269662922\n",
      "24 fimW             count: 444 \tproportion: 0.9977528089887641\n",
      "25 fimY             count: 444 \tproportion: 0.9977528089887641\n",
      "26 fimZ             count: 443 \tproportion: 0.9955056179775281\n",
      "27 gogB             count: 43 \tproportion: 0.09662921348314607\n",
      "28 hilA             count: 445 \tproportion: 1.0\n",
      "29 hilC             count: 444 \tproportion: 0.9977528089887641\n",
      "30 hilD             count: 444 \tproportion: 0.9977528089887641\n",
      "31 iacP             count: 445 \tproportion: 1.0\n",
      "32 iagB             count: 445 \tproportion: 1.0\n",
      "33 invA             count: 445 \tproportion: 1.0\n",
      "34 invB             count: 202 \tproportion: 0.45393258426966293\n",
      "35 invC             count: 445 \tproportion: 1.0\n",
      "36 invE             count: 202 \tproportion: 0.45393258426966293\n",
      "37 invF             count: 277 \tproportion: 0.6224719101123596\n",
      "38 invG             count: 445 \tproportion: 1.0\n",
      "39 invH             count: 445 \tproportion: 1.0\n",
      "40 invI             count: 202 \tproportion: 0.45393258426966293\n",
      "41 invJ             count: 445 \tproportion: 1.0\n",
      "42 lpfA             count: 210 \tproportion: 0.47191011235955055\n",
      "43 lpfB             count: 211 \tproportion: 0.47415730337078654\n",
      "44 lpfC             count: 211 \tproportion: 0.47415730337078654\n",
      "45 lpfD             count: 207 \tproportion: 0.4651685393258427\n",
      "46 lpfE             count: 210 \tproportion: 0.47191011235955055\n",
      "47 mgtB             count: 445 \tproportion: 1.0\n",
      "48 mgtC             count: 445 \tproportion: 1.0\n",
      "49 mig-14           count: 440 \tproportion: 0.9887640449438202\n",
      "50 mig-5            count: 13 \tproportion: 0.029213483146067417\n",
      "51 misL             count: 429 \tproportion: 0.9640449438202248\n",
      "52 orgA             count: 198 \tproportion: 0.4449438202247191\n",
      "53 orgB             count: 445 \tproportion: 1.0\n",
      "54 orgC             count: 333 \tproportion: 0.748314606741573\n",
      "55 pefA             count: 6 \tproportion: 0.01348314606741573\n",
      "56 pefB             count: 6 \tproportion: 0.01348314606741573\n",
      "57 pefC             count: 6 \tproportion: 0.01348314606741573\n",
      "58 pefD             count: 3 \tproportion: 0.006741573033707865\n",
      "59 pegA             count: 152 \tproportion: 0.3415730337078652\n",
      "60 pegB             count: 152 \tproportion: 0.3415730337078652\n",
      "61 pegC             count: 152 \tproportion: 0.3415730337078652\n",
      "62 pegD             count: 152 \tproportion: 0.3415730337078652\n",
      "63 phoP             count: 221 \tproportion: 0.4966292134831461\n",
      "64 phoQ             count: 445 \tproportion: 1.0\n",
      "65 pipB             count: 426 \tproportion: 0.9573033707865168\n",
      "66 pipB2            count: 390 \tproportion: 0.8764044943820225\n",
      "67 pltA             count: 172 \tproportion: 0.3865168539325843\n",
      "68 pltB             count: 162 \tproportion: 0.36404494382022473\n",
      "69 prgH             count: 444 \tproportion: 0.9977528089887641\n",
      "70 prgI             count: 444 \tproportion: 0.9977528089887641\n",
      "71 prgJ             count: 198 \tproportion: 0.4449438202247191\n",
      "72 prgK             count: 445 \tproportion: 1.0\n",
      "73 ratB             count: 403 \tproportion: 0.9056179775280899\n",
      "74 rck              count: 4 \tproportion: 0.008988764044943821\n",
      "75 safA             count: 365 \tproportion: 0.8202247191011236\n",
      "76 safB             count: 366 \tproportion: 0.8224719101123595\n",
      "77 safC             count: 366 \tproportion: 0.8224719101123595\n",
      "78 safD             count: 368 \tproportion: 0.8269662921348314\n",
      "79 sefA             count: 9 \tproportion: 0.020224719101123594\n",
      "80 sefB             count: 8 \tproportion: 0.017977528089887642\n",
      "81 sefC             count: 8 \tproportion: 0.017977528089887642\n",
      "82 sefD             count: 4 \tproportion: 0.008988764044943821\n",
      "83 shdA             count: 402 \tproportion: 0.903370786516854\n",
      "84 sicA             count: 445 \tproportion: 1.0\n",
      "85 sicP             count: 208 \tproportion: 0.46741573033707867\n",
      "86 sifA             count: 444 \tproportion: 0.9977528089887641\n",
      "87 sifB             count: 440 \tproportion: 0.9887640449438202\n",
      "88 sinH             count: 421 \tproportion: 0.946067415730337\n",
      "89 sipA             count: 391 \tproportion: 0.8786516853932584\n",
      "90 sipB             count: 412 \tproportion: 0.9258426966292135\n",
      "91 sipC             count: 396 \tproportion: 0.8898876404494382\n",
      "92 sipD             count: 445 \tproportion: 1.0\n",
      "93 slrP             count: 432 \tproportion: 0.9707865168539326\n",
      "94 sodCI            count: 75 \tproportion: 0.16853932584269662\n",
      "95 sopA             count: 425 \tproportion: 0.9550561797752809\n",
      "96 sopB/sigD        count: 441 \tproportion: 0.9910112359550561\n",
      "97 sopD             count: 441 \tproportion: 0.9910112359550561\n",
      "98 sopD2            count: 421 \tproportion: 0.946067415730337\n",
      "99 sopE             count: 106 \tproportion: 0.23820224719101124\n",
      "100 sopE2            count: 443 \tproportion: 0.9955056179775281\n",
      "101 spaO             count: 204 \tproportion: 0.4584269662921348\n",
      "102 spaP             count: 202 \tproportion: 0.45393258426966293\n",
      "103 spaQ             count: 445 \tproportion: 1.0\n",
      "104 spaR             count: 202 \tproportion: 0.45393258426966293\n",
      "105 spaS             count: 445 \tproportion: 1.0\n",
      "106 spiC/ssaB        count: 445 \tproportion: 1.0\n",
      "107 sprB             count: 444 \tproportion: 0.9977528089887641\n",
      "108 sptP             count: 445 \tproportion: 1.0\n",
      "109 spvA             count: 11 \tproportion: 0.024719101123595506\n",
      "110 spvB             count: 11 \tproportion: 0.024719101123595506\n",
      "111 spvC             count: 17 \tproportion: 0.038202247191011236\n",
      "112 spvD             count: 7 \tproportion: 0.015730337078651686\n",
      "113 spvR             count: 11 \tproportion: 0.024719101123595506\n",
      "114 ssaC             count: 445 \tproportion: 1.0\n",
      "115 ssaD             count: 229 \tproportion: 0.5146067415730337\n",
      "116 ssaE             count: 445 \tproportion: 1.0\n",
      "117 ssaG             count: 239 \tproportion: 0.5370786516853933\n",
      "118 ssaH             count: 445 \tproportion: 1.0\n",
      "119 ssaI             count: 227 \tproportion: 0.5101123595505618\n",
      "120 ssaJ             count: 445 \tproportion: 1.0\n",
      "121 ssaK             count: 231 \tproportion: 0.5191011235955056\n",
      "122 ssaL             count: 445 \tproportion: 1.0\n",
      "123 ssaM             count: 232 \tproportion: 0.5213483146067416\n",
      "124 ssaN             count: 231 \tproportion: 0.5191011235955056\n",
      "125 ssaO             count: 416 \tproportion: 0.9348314606741573\n",
      "126 ssaP             count: 231 \tproportion: 0.5191011235955056\n",
      "127 ssaQ             count: 445 \tproportion: 1.0\n",
      "128 ssaR             count: 445 \tproportion: 1.0\n",
      "129 ssaS             count: 442 \tproportion: 0.9932584269662922\n",
      "130 ssaT             count: 232 \tproportion: 0.5213483146067416\n",
      "131 ssaU             count: 445 \tproportion: 1.0\n",
      "132 ssaV             count: 445 \tproportion: 1.0\n",
      "133 sscA             count: 444 \tproportion: 0.9977528089887641\n",
      "134 sscB             count: 445 \tproportion: 1.0\n",
      "135 sseA             count: 445 \tproportion: 1.0\n",
      "136 sseB             count: 444 \tproportion: 0.9977528089887641\n",
      "137 sseC             count: 445 \tproportion: 1.0\n",
      "138 sseD             count: 440 \tproportion: 0.9887640449438202\n",
      "139 sseE             count: 445 \tproportion: 1.0\n",
      "140 sseF             count: 443 \tproportion: 0.9955056179775281\n",
      "141 sseG             count: 442 \tproportion: 0.9932584269662922\n",
      "142 sseI/srfH        count: 69 \tproportion: 0.1550561797752809\n",
      "143 sseJ             count: 440 \tproportion: 0.9887640449438202\n",
      "144 sseK1            count: 270 \tproportion: 0.6067415730337079\n",
      "145 sseK2            count: 230 \tproportion: 0.5168539325842697\n",
      "146 sseL             count: 439 \tproportion: 0.9865168539325843\n",
      "147 sspH1            count: 194 \tproportion: 0.43595505617977526\n",
      "148 sspH2            count: 267 \tproportion: 0.6\n",
      "149 ssrA             count: 445 \tproportion: 1.0\n",
      "150 ssrB             count: 445 \tproportion: 1.0\n",
      "151 staA             count: 174 \tproportion: 0.39101123595505616\n",
      "152 staB             count: 123 \tproportion: 0.27640449438202247\n",
      "153 staC             count: 180 \tproportion: 0.4044943820224719\n",
      "154 staD             count: 173 \tproportion: 0.3887640449438202\n",
      "155 staE             count: 173 \tproportion: 0.3887640449438202\n",
      "156 staF             count: 173 \tproportion: 0.3887640449438202\n",
      "157 staG             count: 174 \tproportion: 0.39101123595505616\n",
      "158 stbA             count: 242 \tproportion: 0.5438202247191011\n",
      "159 stbB             count: 411 \tproportion: 0.9235955056179775\n",
      "160 stbC             count: 410 \tproportion: 0.9213483146067416\n",
      "161 stbD             count: 410 \tproportion: 0.9213483146067416\n",
      "162 stbE             count: 248 \tproportion: 0.5573033707865168\n",
      "163 stcA             count: 110 \tproportion: 0.24719101123595505\n",
      "164 stcB             count: 175 \tproportion: 0.39325842696629215\n",
      "165 stcC             count: 175 \tproportion: 0.39325842696629215\n",
      "166 stcD             count: 175 \tproportion: 0.39325842696629215\n",
      "167 stdA             count: 430 \tproportion: 0.9662921348314607\n",
      "168 stdB             count: 427 \tproportion: 0.9595505617977528\n",
      "169 stdC             count: 379 \tproportion: 0.851685393258427\n",
      "170 steA             count: 436 \tproportion: 0.9797752808988764\n",
      "171 steB             count: 433 \tproportion: 0.9730337078651685\n",
      "172 steC             count: 433 \tproportion: 0.9730337078651685\n",
      "173 steD             count: 211 \tproportion: 0.47415730337078654\n",
      "174 steE             count: 128 \tproportion: 0.2876404494382023\n",
      "175 steF             count: 211 \tproportion: 0.47415730337078654\n",
      "176 stfA             count: 112 \tproportion: 0.251685393258427\n",
      "177 stfC             count: 263 \tproportion: 0.5910112359550562\n",
      "178 stfD             count: 265 \tproportion: 0.5955056179775281\n",
      "179 stfE             count: 258 \tproportion: 0.5797752808988764\n",
      "180 stfF             count: 112 \tproportion: 0.251685393258427\n",
      "181 stfG             count: 250 \tproportion: 0.5617977528089888\n",
      "182 stgA             count: 12 \tproportion: 0.02696629213483146\n",
      "183 stgB             count: 12 \tproportion: 0.02696629213483146\n",
      "184 stgC             count: 12 \tproportion: 0.02696629213483146\n",
      "185 stgD             count: 12 \tproportion: 0.02696629213483146\n",
      "186 sthA             count: 440 \tproportion: 0.9887640449438202\n",
      "187 sthB             count: 440 \tproportion: 0.9887640449438202\n",
      "188 sthC             count: 440 \tproportion: 0.9887640449438202\n",
      "189 sthD             count: 440 \tproportion: 0.9887640449438202\n",
      "190 sthE             count: 440 \tproportion: 0.9887640449438202\n",
      "191 stiA             count: 358 \tproportion: 0.8044943820224719\n",
      "192 stiB             count: 358 \tproportion: 0.8044943820224719\n",
      "193 stiC             count: 358 \tproportion: 0.8044943820224719\n",
      "194 stiH             count: 357 \tproportion: 0.802247191011236\n",
      "195 stjB             count: 248 \tproportion: 0.5573033707865168\n",
      "196 stjC             count: 249 \tproportion: 0.5595505617977528\n",
      "197 stkA             count: 135 \tproportion: 0.30337078651685395\n",
      "198 stkB             count: 135 \tproportion: 0.30337078651685395\n",
      "199 stkC             count: 135 \tproportion: 0.30337078651685395\n",
      "200 stkD             count: 134 \tproportion: 0.30112359550561796\n",
      "201 stkE             count: 135 \tproportion: 0.30337078651685395\n",
      "202 stkF             count: 135 \tproportion: 0.30337078651685395\n",
      "203 stkG             count: 129 \tproportion: 0.2898876404494382\n",
      "204 tcfA             count: 192 \tproportion: 0.4314606741573034\n",
      "205 tcfB             count: 192 \tproportion: 0.4314606741573034\n",
      "206 tcfC             count: 197 \tproportion: 0.44269662921348313\n",
      "207 tcfD             count: 202 \tproportion: 0.45393258426966293\n",
      "208 tviA             count: 5 \tproportion: 0.011235955056179775\n",
      "209 tviB             count: 5 \tproportion: 0.011235955056179775\n",
      "210 tviC             count: 5 \tproportion: 0.011235955056179775\n",
      "211 tviD             count: 5 \tproportion: 0.011235955056179775\n",
      "212 tviE             count: 2 \tproportion: 0.0044943820224719105\n",
      "213 vexA             count: 5 \tproportion: 0.011235955056179775\n",
      "214 vexB             count: 5 \tproportion: 0.011235955056179775\n",
      "215 vexC             count: 5 \tproportion: 0.011235955056179775\n",
      "216 vexD             count: 5 \tproportion: 0.011235955056179775\n",
      "217 vexE             count: 5 \tproportion: 0.011235955056179775\n",
      "Min: 89 \tMax: 169 \tAvg: 139.50112359550562\n"
     ]
    }
   ],
   "source": [
    "vgenes = sorted([i for i in list(vgc.keys()) if i is not None])\n",
    "for i in vgenes:\n",
    "    try:\n",
    "        print(vgenes.index(i)+1, i, ' '*(15-len(i)), 'count:', vgc[i], '\\tproportion:', vgc[i]/len(go))\n",
    "    except:\n",
    "        continue\n",
    "print('Min:', min([gvgc[i] for i in gvgc.keys()]), '\\tMax:', max([gvgc[i] for i in gvgc.keys()]),'\\tAvg:', \\\n",
    "      sum([gvgc[i] for i in gvgc.keys()])/len(go))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "##Virulence Gene Image - refined a bit\n",
    "##Include adherence operons, effector genes, and toxin genes\n",
    "##pixel dimensions\n",
    "bh = 10 #box height\n",
    "bw = 40 #box width\n",
    "lw = 0 #line width\n",
    "ow = 0 #outline width\n",
    "black = [0,0,0]\n",
    "white = [255,255,255]\n",
    "grey = [100,100,100]\n",
    "\n",
    "#vg class indices in excel file\n",
    "class_inds = [i+1 for i in range(ws.max_column) if ws.cell(column = i+1, row = 1).value is not None]\n",
    "class_inds.append(ws.max_column+2)\n",
    "\n",
    "color_my_lines = True\n",
    "\n",
    "vgc = {}\n",
    "gvgc =  {}\n",
    "for ind in class_inds[:-1]:\n",
    "    class_name = ws.cell(column=ind, row=1).value\n",
    "    outfile = '2017-05-17_VFDB_'+class_name+'_image.png'\n",
    "    start = ind\n",
    "    stop = class_inds[class_inds.index(ind)+1]\n",
    "    \n",
    "    height = ((ws.max_row-3)*(bh+lw))-lw+(2*ow)\n",
    "    width = ((stop-start)*(bw+lw))-lw+(2*ow)\n",
    "\n",
    "    data = np.zeros((height,width,3), np.uint8)\n",
    "\n",
    "    if color_my_lines == True:\n",
    "        color = grey\n",
    "        lxs = [range(ow-lw+((bw+lw)*i),(ow+((bw+lw)*i))) for i in range(1,stop-start)]#line coloring values x axis\n",
    "        lys = [range(ow-lw+((bh+lw)*i),(ow+((bh+lw)*i))) for i in range(1,ws.max_row-3)]#line coloring values y axis\n",
    "        \n",
    "        #white line dividers for different subspecies/clades\n",
    "        lysw = [range(ow-lw+((bh+lw)*(go.index(i)+1)),(ow+((bh+lw)*(go.index(i)+1)))) for i in gos]\n",
    "        \n",
    "        for x in lxs:\n",
    "            for xx in x:\n",
    "                for yy in range(height):\n",
    "                    data[yy,xx]=color\n",
    "        for y in lys:\n",
    "            for yy in y:\n",
    "                for xx in range(width):\n",
    "                    data[yy,xx]=color\n",
    "        for y in lysw:#Dividers for clades\n",
    "            for yy in y:\n",
    "                for xx in range(width):\n",
    "                    data[yy,xx]=white \n",
    "                    \n",
    "        ##color in outlines\n",
    "        for x in range(ow):\n",
    "            for y in range(height):\n",
    "                data[y,x]=color\n",
    "        for x in range(width-ow,width):\n",
    "            for y in range(height):\n",
    "                data[y,x]=color\n",
    "        for y in range(ow):\n",
    "            for x in range(width):\n",
    "                data[y,x]=color\n",
    "        for y in range(height-ow,height):\n",
    "            for x in range(width):\n",
    "                data[y,x]=color\n",
    "    \n",
    "    for i in range(stop-start):\n",
    "        gene = ws.cell(column=i+start,row=3).value\n",
    "        geneclass = ws.cell(column=i+start, row=2).value\n",
    "        vgc[gene] = 0\n",
    "        \n",
    "        ##if new class and not the first, color in lines to the left\n",
    "        if (i != 0) and (geneclass != None):\n",
    "            for xx in range(ow-lw+((bw+lw)*i),(ow+((bw+lw)*i))):\n",
    "                for yy in range(height):\n",
    "                    data[yy,xx]=white\n",
    "            \n",
    "        \n",
    "        for j in range(ws.max_row-3):\n",
    "            genome = ws.cell(column=1,row=j+4).value\n",
    "            if genome not in gvgc.keys():\n",
    "                gvgc[genome] = 0\n",
    "            if ws.cell(column=i+start,row=j+4).value is not None:\n",
    "                pct = float(ws.cell(column=i+start,row=j+4).value)\n",
    "                red = round(255*((100-pct)/20))\n",
    "                if red > 255:\n",
    "                    red == 255\n",
    "                green = 255\n",
    "                color =[red, green, 0]\n",
    "                for h in range(ow+(go.index(genome)*(bh+lw)),ow+(go.index(genome)*(bh+lw))+bh):\n",
    "                    for w in range(ow+(i*(bw+lw)),ow+(i*(bw+lw))+bw):\n",
    "                        data[h,w] = color\n",
    "                vgc[gene] += 1\n",
    "                gvgc[genome] += 1\n",
    "    img = smp.toimage(data).save(outfile)\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
